components/director/internal/healthz/indicator.go

```diff
@@ -0,0 +1,146 @@
+package healthz
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+)
+
+//go:generate mockery --name=Indicator --output=automock --outpkg=automock --case=underscore
+type Indicator interface {
+	Name() string
+	Configure(IndicatorConfig)
+	Run(ctx context.Context) error
+	Status() Status
+}
+
+//go:generate mockery --name=Status --output=automock --outpkg=automock --case=underscore
+type Status interface {
+	Error() error
+	Details() string
+}
+
+type IndicatorFunc func(ctx context.Context) Status
+
+// Implements Status interface
+type status struct {
+	error   error
+	details string
+}
+
+// Error returns status error
+func (s *status) Error() error {
+	return s.error
+}
+
+// Details returns status details
+func (s *status) Details() string {
+	return s.details
+}
+
+// Implements Indicator interface
+type indicator struct {
+	name string
+
+	interval     time.Duration
+	timeout      time.Duration
+	initialDelay time.Duration
+	threshold    int
+
+	indicatorFunc IndicatorFunc
+	status        Status
+	statusLock    sync.Mutex
+	failureCount  int
+}
+
+// NewIndicator returns new indicator from the provided IndicatorConfig and IndicatorFunc
+func NewIndicator(name string, indicatorFunc IndicatorFunc) Indicator {
+	return &indicator{
+		name:          name,
+		indicatorFunc: indicatorFunc,
+		status:        &status{details: ""},
+		statusLock:    sync.Mutex{},
+		failureCount:  0,
+	}
+}
+
+// Name returns indicator name
+func (i *indicator) Name() string {
+	return i.name
+}
+
+// Configure sets indicator config based on IndicatorConfig
+func (i *indicator) Configure(cfg IndicatorConfig) {
+	i.interval = cfg.Interval
+	i.timeout = cfg.Timeout
+	i.initialDelay = cfg.InitialDelay
+	i.threshold = cfg.Threshold
+}
+
+// Run starts the periodic indicator checks
+func (i *indicator) Run(ctx context.Context) error {
+	if i.interval <= 0 {
+		return errors.New("indicator interval could not be <= 0")
+	}
+	if i.timeout <= 0 {
+		return errors.New("indicator timeout could not be <= 0")
+	}
+	if i.initialDelay < 0 {
+		return errors.New("indicator initial delay could not be < 0")
+	}
+	if i.threshold < 0 {
+		return errors.New("indicator threshold could not be < 0")
+	}
```

Comment:
Can this snippet be part of the configure method?

Reply:
I moved it to the `New` health function. This way we could pipe multiple `RegisterIndicator` calls.